docs/examples.rst

```diff
@@ -0,0 +1,310 @@
+
+Note that MNIST is such a small dataset that yogadl is not going to outperform
+any example that treats MNIST as an in-memory dataset.
+
+.. code:: python
+
+    import math
+    import os
+    import tensorflow as tf
+    import tensorflow_datasets as tfds
+    import yogadl
+    import yogadl.tensorflow
+    import yogadl.storage
+
+    BATCH_SIZE = 32
+
+    # Configure the yogadl storage.
+    storage_path = "/tmp/yogadl_cache"
+    os.makedirs(storage_path, exist_ok=True)
+    lfs_config = yogadl.storage.LFSConfigurations(storage_path)
+    storage = yogadl.storage.LFSStorage(lfs_config)
+
+    @storage.cacheable("mnist", "1.0")
+    def make_data():
+        mnist = tfds.image.MNIST()
+        mnist.download_and_prepare()
+        dataset = mnist.as_dataset(as_supervised=True)["train"]
+
+        # Apply dataset transformations from the TensorFlow docs:
+        # (https://www.tensorflow.org/datasets/keras_example)
+
+        def normalize_img(image, label):
+            """Normalizes images: `uint8` -> `float32`."""
+            return tf.cast(image, tf.float32) / 255., label
+
+        return dataset.map(normalize_img)
+
+    # Get the DataRef from the storage via the decorated function.
+    dataref = make_data()
+
+    # Stream the dataset starting halfway through it.
+    num_batches = math.ceil(len(dataref) / BATCH_SIZE)
+    batches_to_skip = num_batches // 2
+    records_to_skip = batches_to_skip * BATCH_SIZE
+    stream = dataref.stream(
+        start_offset=records_to_skip, shuffle=True, shuffle_seed=777
+    )
+
+    # Convert the stream to a tf.data.Dataset object.
+    dataset = yogadl.tensorflow.make_tf_dataset(stream)
+
+    # Apply normal data augmentation and prefetch steps.
+    dataset = dataset.batch(BATCH_SIZE)
+    dataset = dataset.prefetch(tf.data.experimental.AUTOTUNE)
+
+    # Model is straight from the TensorFlow docs:
+    # https://www.tensorflow.org/datasets/keras_example
+    model = tf.keras.models.Sequential([
+      tf.keras.layers.Flatten(input_shape=(28, 28, 1)),
+      tf.keras.layers.Dense(128,activation='relu'),
+      tf.keras.layers.Dense(10, activation='softmax')
+    ])
+    model.compile(
+        loss='sparse_categorical_crossentropy',
+        optimizer=tf.keras.optimizers.Adam(0.001),
+        metrics=['accuracy'],
+    )
+    model.fit(dataset)
+
+Advanced Use Case: Distributed Training
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Sharding a dataset for use with distributed training is easy. If you are using
+Horovod for distributed training, you only need to alter the arguments of your
+call to ``DataRef.stream()``.
+
+.. code:: python
+
+    import horovod.tensorflow as hvd
+
+    ...
+
+    stream = dataref.stream(
+        shard_rank=hvd.rank(), num_shards=hvd.size()
+    )
+
+Advanced Use Case: Custom DataRef Objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Suppose you would like to write a custom dataset for use with a platform that
```

Comment:
? If you have an advanced use case, like generating data on an external machine and streaming it
to another machine for training, that isn't currently covered by `yogadl` you can implement a custom `yogadl.DataRef`. ...